pkg/convert/pprof/streaming/parser_streaming.go

```diff
@@ -213,20 +219,25 @@ func (p *VTStreamingParser) addStackFrame(l *line) error {
 	if !ok {
 		return nil
 	}
+	var frame []byte
 	switch p.Formatter {
 	case StackFrameFormatterRuby:
 		pFuncName := p.strings[f.name]
 		pFileName := p.strings[f.filename]
-		frame := []byte(fmt.Sprintf("%s:%d - %s",
+		frame = []byte(fmt.Sprintf("%s:%d - %s",
 			p.profile[(pFileName>>32):(pFileName&0xffffffff)],
 			l.line,
 			p.profile[(pFuncName>>32):(pFuncName&0xffffffff)]))
-		p.tmpSample.tmpStack = append(p.tmpSample.tmpStack, frame)
 	default:
 	case StackFrameFormatterGo:
 		pFuncName := p.strings[f.name]
-		frame := p.profile[(pFuncName >> 32):(pFuncName & 0xffffffff)]
-		p.tmpSample.tmpStack = append(p.tmpSample.tmpStack, frame)
+		frame = p.profile[(pFuncName >> 32):(pFuncName & 0xffffffff)]
+	}
+	pSample := &p.tmpSample
+	if len(pSample.tmpStack) < cap(pSample.tmpStack) {
+		pSample.tmpStack = append(pSample.tmpStack, frame)
+	} else {
+		pSample.tmpStack = arenahelper.AppendA(pSample.tmpStack, frame, p.arena)
```

Comment:
I saw this many times in the PR: I must me missing something but should not `arenahelper` be responsible of that?

Reply:
This is an obscure optimization attempt.
arenahelper.AppendA is too complex for inlining. 
and with this manual check it becomes just a bunch of instructions like `slice.data[slice.len++]=newVal` with no function call.

Reply:
Out of curiosity: does PGO help in this particular case? What is the impact of the optimisation?

Reply:
I've just tried to do pgo thing on a trivial program but could not make compiler inline the appendA function and no performance improvements found after pgo applied

I tried it on the following program:

https://gist.github.com/korniltsev/189f66aa2f2ca3706e2765ef884548f4

```
go build  . 
./main
go build -pgo=cpu.pprof .
./main
```


maybe Im missing smth or doing something wrong, did not spend a lot of time on it

Reply:
I wrote a quick bench to demonstrate impact
````
package main

import (
	"arenaspg/arenahelper"
	"fmt"
	"testing"
)

var testcases = []int{16, 64, 256}

func BenchmarkOp(b *testing.B) {
	for _, n := range testcases {
		b.Run(fmt.Sprintf("n=%d", n), func(b *testing.B) {
			a := arenahelper.NewArenaWrapper()
			s := arenahelper.MakeSlice[int](a, 0, n)
			for i := 0; i < b.N; i++ {
				s = s[:0]
				for i := 0; i < n; i++ {
					if len(s) < cap(s) {
						s = append(s, i)
					} else {
						s = arenahelper.AppendA(s, i, a)
					}
				}
				bh(s)
			}
		})
	}
}

func BenchmarkNoOP(b *testing.B) {
	for _, n := range testcases {
		b.Run(fmt.Sprintf("n=%d", n), func(b *testing.B) {
			a := arenahelper.NewArenaWrapper()
			s := arenahelper.MakeSlice[int](a, 0, n)
			for i := 0; i < b.N; i++ {
				s = s[:0]
				for i := 0; i < n; i++ {
					s = arenahelper.AppendA(s, i, a)
				}
				bh(s)
			}
		})
	}
}
````


```
goos: darwin
goarch: arm64
pkg: arenaspg
BenchmarkOp
BenchmarkOp/n=16
BenchmarkOp/n=16-8         	183499448	         6.537 ns/op
BenchmarkOp/n=64
BenchmarkOp/n=64-8         	53859661	        22.25 ns/op
BenchmarkOp/n=256
BenchmarkOp/n=256-8        	12262503	        97.54 ns/op
BenchmarkNoOP
BenchmarkNoOP/n=16
BenchmarkNoOP/n=16-8       	47982326	        24.97 ns/op
BenchmarkNoOP/n=64
BenchmarkNoOP/n=64-8       	 9093646	       131.4 ns/op
BenchmarkNoOP/n=256
BenchmarkNoOP/n=256-8      	 2377095	       501.8 ns/op

```

Reply:
haha I think pgo does not apply to generics either intentionally or because of a bug
pprof reports it as  `arenahelper.AppendA[...]` and there is only a `arenahelper.AppendA[go.shape.int]` in call graph

if I manually create a non generic copy of AppendA - AppendInt - it is inlined with pgo

Reply:
Hi,
A little belated advice
It seems splitting method into fast and slow paths should solve inlining problem

```
func AppendA[T any](data []T, v T, a ArenaWrapper) []T {
	if len(data) < cap(data) {
		return append(data, v)
	}

	return appendASlow(data, v, a)
}

func appendASlow[T any](data []T, v T, a ArenaWrapper) []T {
	if a == nil { // move me to the AppendA method when the compiler gets smarter
		return append(data, v)
	}

	c := 2 * len(data)
	if c == 0 {
		c = 1
	}
	newData := arena.MakeSlice[T](a, len(data)+1, c)
	copy(newData, data)
	data = newData
	data[len(data)-1] = v
	return data
}
```

Reply:
That's likely true.
Forthermore updating golang is likely to help. The PR was written at the time of golang 1.20 and there were fixes (using fully quallified generics name in profiles instead of `[...]`)
Note: this code is completely dead and not used right now (was used before grafana merge) .